feat: Mounting resources should not cause workspace restart

tolusha opened the pull request:
### What does this PR do?
 This PR prevents automatic workspace restarts when mounting resources (ConfigMaps, Secrets, PVC) by introducing a new `controller.devfile.io/mount-on-start` attribute.

### What issues does this PR fix or reference?
https://github.com/eclipse-che/che/issues/23553

### Is it tested? How?
Clean namespace before each scenario.

<details>
  <summary>Scenario 1</summary>

1. Start a workspace
2. Create a cm, secret and pvc, ensure workspace is not restarted
```
oc apply -f - <<EOF             
apiVersion: v1  
data:
  CM: data
kind: ConfigMap
metadata:
  annotations:
    controller.devfile.io/mount-as: env
    controller.devfile.io/mount-on-start: "true"
  labels:
    controller.devfile.io/mount-to-devworkspace: "true"
    controller.devfile.io/watch-configmap: "true"
  name: test
---
apiVersion: v1  
stringData:
  SECRET: data
kind: Secret
metadata:
  annotations:
    controller.devfile.io/mount-as: file
    controller.devfile.io/mount-on-start: "true"
    controller.devfile.io/mount-path: "/tmp/secret"
  labels:
    controller.devfile.io/mount-to-devworkspace: "true"
    controller.devfile.io/watch-secret: "true"
  name: test
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-pvc
  labels:
    controller.devfile.io/mount-to-devworkspace: 'true'
  annotations:
    controller.devfile.io/mount-path: /tmp/pvc
    controller.devfile.io/mount-on-start: "true"
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi
  volumeMode: Filesystem
EOF
```
3. Restart a workspace, ensure that resources are mounted
</details>


<details>
  <summary>Scenario 2</summary>

1. Start a workspace
2. Create a git configurations, ensure workspace is not restarted
```
oc apply -f - <<EOF    
kind: Secret
apiVersion: v1
metadata:
  name: test-git-credentials-secret
  labels:
    controller.devfile.io/git-credential: 'true'
    controller.devfile.io/watch-secret: 'true'
  annotations:
    controller.devfile.io/mount-on-start: "true"    
type: Opaque
stringData:
  credentials: https://test:test@test
EOF
```
3. Create one more secret
```
oc apply -f - <<EOF    
kind: Secret
apiVersion: v1
metadata:
  name: test-git-credentials-secret-2
  labels:
    controller.devfile.io/git-credential: 'true'
    controller.devfile.io/watch-secret: 'true'
type: Opaque
stringData:
  credentials: https://test2:test2@test2
EOF
```  
4. Ensure workspace is restarted and all credentials are merged `/.git-credentials/credentials`
</details>

<details>
  <summary>Scenario 3</summary>

1. Start a workspace
2. Create a git configurations, ensure workspace is not restarted.
```
oc apply -f - <<EOF    
apiVersion: v1  
data:
  certificate: CERT
kind: ConfigMap
metadata:
  annotations:
    controller.devfile.io/mount-on-start: "true"
  labels:
    controller.devfile.io/git-tls-credential: "true"
    controller.devfile.io/watch-configmap: 'true'
  name: test-git-config
EOF
```
3. Rewrite the git configuration, ensure workspace is restarted. Configuration exists in `/etc/gitconfig`
```
oc apply -f - <<EOF    
apiVersion: v1  
data:
  certificate: CERT
kind: ConfigMap
metadata:
  labels:
    controller.devfile.io/git-tls-credential: "true"
    controller.devfile.io/watch-configmap: 'true'
  name: test-git-config
EOF
```
</details>


<details>
  <summary>Scenario 4</summary>

1. Start a workspace
2. Create resources while workspace is starting
```
oc apply -f - <<EOF             
apiVersion: v1  
data:
  CM: data
kind: ConfigMap
metadata:
  annotations:
    controller.devfile.io/mount-as: env
    controller.devfile.io/mount-on-start: "true"
  labels:
    controller.devfile.io/mount-to-devworkspace: "true"
    controller.devfile.io/watch-configmap: "true"
  name: test
---
apiVersion: v1  
stringData:
  SECRET: data
kind: Secret
metadata:
  annotations:
    controller.devfile.io/mount-as: file
    controller.devfile.io/mount-on-start: "true"
    controller.devfile.io/mount-path: "/tmp/secret"
  labels:
    controller.devfile.io/mount-to-devworkspace: "true"
    controller.devfile.io/watch-secret: "true"
  name: test
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-pvc
  labels:
    controller.devfile.io/mount-to-devworkspace: 'true'
  annotations:
    controller.devfile.io/mount-path: /tmp/pvc
    controller.devfile.io/mount-on-start: "true"
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi
  volumeMode: Filesystem
EOF
```
3. Ensure resources are mounted

</details>

### PR Checklist

- [x] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [x] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [x] `v8-che-happy-path`: Happy path for verification integration with Che


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added controller.devfile.io/mount-on-start annotation to mount resources only at workspace start; avoids unwanted restarts when resources are created while a workspace is running.
  * Automount now respects current workspace running state and will only add mounts that match the running deployment.

* **Documentation**
  * Updated docs with mount-on-start guidance and YAML examples.

* **Tests**
  * Added unit and controller regression tests covering mount-on-start behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tolusha commented on pkg/provision/automount/gitconfig.go:
```diff
@@ -92,8 +97,13 @@ func getGitResources(api sync.ClusterAPI, namespace string) (credentialSecrets [
+	for _, configmap := range configmapList.Items {
+		// Skip mounting if mount-on-start-only is set to "true" and workspace has been already started
+		mountOnStartOnly := configmap.Annotations[constants.MountOnStartOnlyAttribute] == "true"
+		if isWorkspaceStarted && mountOnStartOnly {
```
Agree

dkwon17 commented:
I haven't tried, but is it possible to ignore configmaps/secrets that have the `controller.devfile.io/mount-on-start-only` attribute in `automountWatcher` to avoid reconciling running workspaces? 

https://github.com/devfile/devworkspace-operator/blob/b61eaed8a7ca6c5f94e983c528ba6d65a7d35227/controllers/workspace/devworkspace_controller.go#L727

tolusha committed:
chore: creating automount objects should not trigger workspace to restart

Signed-off-by: Anatolii Bazko <abazko@redhat.com>

tolusha committed:
chore: creating automount objects should not trigger workspace to restart

Signed-off-by: Anatolii Bazko <abazko@redhat.com>

rohanKanojia commented on docs/additional-configuration.adoc:
```diff
@@ -166,6 +166,17 @@ When "file" is used, the configmap is mounted as a directory within the workspac
 
+* `controller.devfile.io/mount-on-start`: when set to `"true"`, the resource will only be mounted when a workspace starts. If the resource is created while a workspace is already running, it will not be automatically mounted until the workspace is restarted. This prevents unwanted workspace restarts caused by newly created automount resources. This annotation can be applied to configmaps, secrets ...
++
+For git credential secrets (labelled with `controller.devfile.io/git-credential`) and git TLS configmaps (labelled with `controller.devfile.io/git-tls-credential`), the annotation works similarly but applies collectively: since all git credentials are merged into a single mounted secret, if at least one git credential secret (or TLS configmap) lacks the `controller.devfile.io/mount-on-start` anno ...
```
```suggestion
For git credential secrets (labelled with `controller.devfile.io/git-credential`) and git TLS configmaps (labelled with `controller.devfile.io/git-tls-credential`), the annotation is evaluated across all git credential resources as a group, not individually. Since all git credentials are merged into a single mounted secret, if at least one git credential secret (or TLS configmap) lacks the `controller.devfile.io/mount-on-start` annotation, all git credentials (or TLS configmaps) will be mounted, including those marked with `mount-on-start`.
```

rohanKanojia replied:
Maybe we can add a NOTE block:
```
[NOTE]
====
This annotation must be applied consistently across all git credential resources. If it is missing from any resource, it will not be honored for any of them.
====
```

tolusha replied:
I will apply the suggestion, but I don’t think it makes sense to add the note.

rohanKanojia replied:
@tolusha: Okay, np problem. It was just a suggestion

rohanKanojia commented on pkg/provision/automount/gitconfig_test.go:
```diff
@@ -218,6 +220,208 @@ func TestTwoConfigMapWithBothMissingHost(t *testing.T) {
+	}
+}
+
+func TestMountGitCredentialWhenMixedMountOnStartSecrets(t *testing.T) {
```
Does it make sense to add a similar Mixed mount test for git TLS config, too, i.e., `TestMountGitTLSConfigWhenMixedMountOnStartConfigMaps`?

tolusha replied:
Sure

rohanKanojia commented on pkg/provision/automount/common_test.go:
```diff
@@ -407,3 +409,417 @@ func loadTestCaseOrPanic(t *testing.T, testPath string) testCase {
+	assert.Equal(t, common.AutoMountPVCVolumeName("test-pvc"), testPodAdditions.Volumes[0].Name)
+}
+
+func TestShouldNotMountConfigMapWithMountOnStartWhenRunningAndNotInDeployment(t *testing.T) {
```
This test looks very similar to `TestShouldNotMountConfigMapWithMountOnStartIfWorkspaceStarted`.

Just wondering if there's a subtle difference I'm missing.

tolusha replied:
Basically the same test.

rohanKanojia reviewed (commented):
I've tested all scenarios and can confirm they work as expected ✅ . Just minor nitpicks

dkwon17 commented:
I need to do more testing and reviewing, but scenario 3 is not working as expected for me. When I follow step 3:  `Rewrite the git configuration, ensure workspace is restarted ...` I notice that my running workspace is not restarted, any ideas?

tolusha commented:
@dkwon17 
Sorry, I forgot to mention.
You need a clean namespace for every scenario.

tolusha commented:
/generate-che-doc

Documentation PR created: https://github.com/eclipse-che/che-docs/pull/3076

rohanKanojia commented:
@dkwon17 : Hello, I tried again Scenario 3. In my case workspace did restart on applying the ConfigMap a second time for rewriting git configuration

https://github.com/user-attachments/assets/9b2c2b3c-3019-409e-bda7-63d14e5e0f8d

akurinnoy commented on pkg/provision/automount/gitconfig.go:
```diff
@@ -92,8 +105,12 @@ func getGitResources(api sync.ClusterAPI, namespace string) (credentialSecrets [
 	var configmaps []corev1.ConfigMap
-	if len(configmapList.Items) > 0 {
-		configmaps = configmapList.Items
+	// When git credentials are present, the gitconfig ConfigMap must be created
```
This collective behavior is correct and the comment above explains the intent well. However, there's no test that exercises the cross-resource-type scenario specifically - e.g., one git credential secret WITHOUT `mount-on-start` + one TLS configmap WITH `mount-on-start` on a running workspace.

`TestMountGitCredentialWhenMixedMountOnStartSecrets` covers same-type mixing. A similar cross-type test would ensure the `len(secrets) > 0 ||` short-circuit is exercised under test.

akurinnoy commented on pkg/provision/automount/gitconfig.go:
```diff
@@ -139,3 +156,51 @@ func cleanupGitConfig(api sync.ClusterAPI, namespace string) error {
+	return false
+}
+
+func allItemsMountOnStart[T any](objs []T) bool {
```
nit: The `[T any]` constraint is the weakest possible - a caller could pass `[]string` and silently get `true` back (all items would fail the type assertion and be `continue`d). This is a known Go generics limitation since `*T implements k8sclient.Object` can't be expressed as a type constraint.

Consider adding a doc comment noting the implicit requirement:

```go
// allItemsMountOnStart returns true if every item in the slice has the mount-on-start annotation.
// T must implement k8sclient.Object via pointer receiver; non-conforming items are silently skipped.
```

akurinnoy commented on pkg/provision/workspace/deployment.go:
```diff
@@ -396,3 +396,18 @@ func getAdditionalDeploymentAnnotations(workspace *common.DevWorkspaceWithConfig
 	return annotations, nil
 }
+
+func GetClusterDeployment(workspace *common.DevWorkspaceWithConfig, clusterAPI sync.ClusterAPI) (*appsv1.Deployment, error) {
```
nit: This function currently has 0% test coverage per Codecov. While it's straightforward, a test verifying the `IsNotFound -> (nil, nil)` behavior would catch regressions if the error handling changes. Low priority since the logic is minimal.

akurinnoy commented on pkg/provision/automount/gitconfig_test.go:
```diff
@@ -218,6 +220,208 @@ func TestTwoConfigMapWithBothMissingHost(t *testing.T) {
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{Name: "test-container"}},
+					Volumes:    []corev1.Volume{{Name: constants.GitCredentialsMergedSecretName}},
```
This seeds the deployment volume with the raw constant `constants.GitCredentialsMergedSecretName`, but production code constructs volume names via `common.AutoMountSecretVolumeName(...)`. Same issue at line 377 with `constants.GitCredentialsConfigMapName`.

Currently the helper functions are identity functions, but if they ever transform names, these tests would pass falsely without exercising the "already exists in deployment" branch. The PVC test in `common_test.go` correctly uses `common.AutoMountPVCVolumeName("test-pvc")`.

Consider:
```suggestion
					Volumes:    []corev1.Volume{{Name: common.AutoMountSecretVolumeName(constants.GitCredentialsMergedSecretName)}},
```

akurinnoy commented on pkg/provision/automount/common_test.go:
```diff
@@ -407,3 +409,417 @@ func loadTestCaseOrPanic(t *testing.T, testPath string) testCase {
+func TestShouldNotMountConfigMapWithMountOnStartWhenRunningAndNotInDeployment(t *testing.T) {
+	testAPI := sync.ClusterAPI{
+		Client: fake.NewClientBuilder().WithObjects(mountOnStartConfigMapAsFile()).Build(),
+	}
```
This test is functionally identical to `TestShouldNotMountConfigMapWithMountOnStartIfWorkspaceStarted` above - same input, same assertions, same `emptyDeployment()`.

Same pattern repeats for secrets (line 413 vs the secret equivalent further down) and PVCs. Consider removing one of each pair or repurposing them to cover a genuinely different scenario - e.g., a resource without the `mount-on-start` annotation on a running workspace (verifying the `!isMountOnStart -> return true` branch in `isAllowedToMount`).

akurinnoy commented on pkg/provision/automount/common_test.go:
```diff
@@ -407,3 +409,417 @@ func loadTestCaseOrPanic(t *testing.T, testPath string) testCase {
+	assert.Equal(t, common.AutoMountConfigMapVolumeName("test-cm"), testPodAdditions.Containers[0].EnvFrom[0].ConfigMapRef.Name)
+}
+
+func TestMountOnStartConfigMapAsFileAllowedWhenVolumeExistsInDeployment(t *testing.T) {
```
nit: `AutoMountConfigMapVolumeName` generates a *volume* name, not an EnvFrom ref name. The assertion passes because the function is currently an identity function and the configmap name happens to match. But semantically, the EnvFrom ref name should be compared to the ConfigMap's object name directly (`"test-cm"`), not to a volume name helper output.

Same pattern at the secret equivalent further down.

akurinnoy reviewed (approved):
Just a few minor nitpicks.

akurinnoy commented:
A couple of general suggestions:

1. **Debug logging for skipped resources** - When a mount-on-start resource is skipped during reconciliation, there's no log entry or event. Consider adding a debug-level log in the `isAllowedToMount` path when a resource is skipped, e.g.:
   ```go
   log.V(4).Info("Skipping mount-on-start resource - not present in running deployment", "name", obj.GetName())
   ```
   This would help operators troubleshoot why a resource is not visible in a running workspace.

2. **Documentation clarification** - `docs/additional-configuration.adoc` explains what happens when the `mount-on-start` annotation is present, but doesn't describe what happens when it's removed from an already-mounted resource. Consider adding a sentence like: "Removing the annotation from an already-mounted resource does not unmount it from a running workspace - the change takes effect on the next workspace restart."

tolusha committed:
fixup

Signed-off-by: Anatolii Bazko <abazko@redhat.com>

tolusha commented:
@rohanKanojia 
@akurinnoy 

Pls review, I've addressed most of remarks.
Added and improved tests

> 1. Debug logging for skipped resources

I can’t add logging here, as I’m concerned it may lead to excessive log spam.

> 2. Documentation clarification 

Unfortunately, this doesn’t seem meaningful or applicable in this context.

dkwon17 commented:
@tolusha how about logging with `log.V(1).Info`, so that DWO by default will not display the log?

tolusha committed:
chore: Add logging

Signed-off-by: Anatolii Bazko <abazko@redhat.com>

tolusha commented:
/retest

tolusha commented:
/retest
